{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Processing ATAC-seq Data into Base-Level Data\n",
    "### Goal  \n",
    "- The `goal` of this procedure is to transform raw single-nucleus ATAC-seq (snATAC-seq) data from `multiome` libraries into `base-level` chromatin accessibility data for downstream analysis. \n",
    "- The purpose is to work with finer resolution data (at the base level), rather than relying on peak calling, which groups accessible chromatin regions. \n",
    "- Base-level data offers a more `detailed view` of chromatin accessibility across individual bases, which can be critical for certain models like fine-mapping or multiomic studies, such as fitting the `mfsusie`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data\n",
    "- The `ROSMAP` single-nucleus ATAC-seq (`snATAC-seq`) is from the `10x Genomics` platform, specifically from the `multiome` libraries. \n",
    "- **Samples**: It containes` > 200 samples`.\n",
    "- **Path**: `/mnt/mfs/ctcn/datasets/rosmap/snmultiome/dlpfc/batch1`    \n",
    "    - fragment files: `/mnt/mfs/ctcn/datasets/rosmap/snmultiome/dlpfc/batch1/cellRanger/lirary_id/outs/atac_fragments.tsv.gz`   \n",
    "    - sample data: `/mnt/mfs/ctcn/datasets/rosmap/snmultiome/dlpfc/batch1/demuxlet/scripts/SampleSheet.csv`   \n",
    "    - library QCed Seurat object: `/mnt/mfs/ctcn/datasets/rosmap/snmultiome/dlpfc/batch1/atac_qc/analysis/step1_chromatin_assay_QC/snMultiome_chromatin_assay_libraryidtoid.rds`    \n",
    "    - cell type annotated Seurat object: `/mnt/mfs/ctcn/datasets/rosmap/snmultiome/dlpfc/batch1/atac_qc/analysis/step3_postQC/'cell_type'/cell_type.rds`   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Overview of the Process:\n",
    "`Procedure`: fragnent QC(library) -> cell type annotation(library) ->  split by cell type and sample(6 cell type *240 samples) ->  Calculate coverage(6*240 bedgraph files) -> Merge all sample bedgraph into one for the sample cell type(6 bedgraph files)  ->  QC and get base by sample count matrix of each cell type \n",
    "\n",
    "`Input Data`: The starting data consists of `fragment` files generated from snATAC-seq. Each fragment file contains records of accessible chromatin fragments with the following information:    \n",
    "- Chromosome\n",
    "- Start and end positions of the fragment\n",
    "- Barcode identifying the cell/nucleus\n",
    "- Read count information\n",
    "\n",
    "Data Structure:   \n",
    "```bash\n",
    "## Fragment file format:\n",
    "chr start end barcode reads\n",
    "chr1 10073 10210 AGTAGGATCCCGTTAC-1 1\n",
    "chr1 10073 10309 TCGCGCACAGCAAATA-1 1\n",
    "chr1 10079 10261 ACATCAATCCGCCAAA-1 2\n",
    "chr1 10084 10204 TCATTGTTCCGCCTAT-1 1\n",
    "chr1 10085 10303 ATTGGCTAGCGCCTAA-1 1\n",
    "chr1 10085 10340 GAACTTATCCCTGATC-1 1\n",
    "```\n",
    "This format shows chromatin fragments for specific cells, identified by barcodes. The coordinates of the start and end positions represent where the chromatin is accessible.\n",
    "\n",
    "`Purpose of Coverage`:\n",
    "- Coverage is a measure of how many fragments (or reads) map to a specific base across all cells. It helps identify regions of high or low chromatin accessibility at the single-base resolution.  \n",
    "- Tools like `bedtools`  are typically used to calculate coverage from fragment files.    \n",
    "For example, calculating coverage will result in:   \n",
    "\n",
    "```bash\n",
    "| Base Position    | Sample1 |\n",
    "|------------------|---------|\n",
    "| chr1:100000      | 1       |\n",
    "| chr1:100001      | 0       |\n",
    "| chr1:100002      | 3       |\n",
    "```\n",
    "This output shows how many fragments overlap each base, giving a base-by-base accessibility profile."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Procedure for Processing ATAC-seq Data to Base-Level Data\n",
    "\n",
    "61 libraries(~ 4 samples/library) are processed in this procedure. The process involves several steps to transform raw fragment data into base-level chromatin accessibility data for each cell type. The steps are as follows:\n",
    "\n",
    "**1) QC Fragment Files**:   \n",
    "\n",
    "- Start with raw fragment files, which contain chromatin fragments with information such as chromosome positions, start and end points, and associated barcodes.   \n",
    "- Apply quality control (QC) by matching filtered QCed barcodes (e.g., ncell, nfeature) from Seurat object metadata with the fragments. This ensures only high-quality cells are retained for further analysis.  \n",
    "\n",
    "**2) Subset by Cell Type**:\n",
    "\n",
    "- Use an annotation file (another Seurat object) to map barcodes to specific cell types.   \n",
    "- Subset the fragments per cell type, ensuring each fragment is correctly assigned to its corresponding cell type based on the barcodes.  \n",
    "\n",
    "**3) Subset by Sample**:   \n",
    "\n",
    "- Further subset the cell-type-specific fragments by sample ID.   \n",
    "- Each sample within the cell type is separated out to individual files for more granular analysis. \n",
    "\n",
    "**4) Run Coverage on Per-Sample Files**:   \n",
    "\n",
    "- For each sample-specific fragment file, calculate base-level chromatin accessibility using bedtools coverage.\n",
    "- `bedtools -bg`: for BedGraph output, which is more memory-efficient (1-2 mins/each).\n",
    "- This produces per-sample `BedGraph` files, showing the chromatin accessibility at each base position.\n",
    "\n",
    "**5) Merge BedGraph Files by Cell Type:**\n",
    "\n",
    "- Merge the BedGraph files from different samples within the same cell type using `bedtools unionbedg`.\n",
    "- This step combines data from multiple samples of the same cell type into a single base-by-sample matrix, ready for analysis.\n",
    "\n",
    "**6) QC and get Base-by-Sample Matrix:**\n",
    "\n",
    "This pseudobulk-level matrix can undergo QC to filter out low-quality samples, ensuring reliable downstream analysis. The final matrix contains base-level chromatin accessibility for each sample within a given cell type.\n",
    "\n",
    "\n",
    "```bash\n",
    "## Final Output: Base-by-Sample Matrix\n",
    "\n",
    "| Base Position    | Sample1 | Sample2 | Sample3 | ... |\n",
    "|------------------|---------|---------|---------|-----|\n",
    "| chr1:100000      | 10      | 5       | 8       | ... |\n",
    "| chr1:100001      | 0       | 1       | 3       | ... |\n",
    "| chr1:100002      | 12      | 0       | 15      | ... |\n",
    "| chr1:100003      | 7       | 4       | 10      | ... |\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### BedGraph Data Format and Explanation\n",
    "- The `BedGraph` format is more memory-efficient. Unlike `base-level` data, which stores coverage at every base, BedGraph condenses `consecutive` positions with the `same` coverage value into a single `interval`, reducing redundancy. \n",
    "- This dramatically reduces memory usage, making it a better format for storing and processing large genomic datasets at scale. This format is particularly helpful during `QC`, and we can convert the interval data into base-level data for final analysis if needed.\n",
    "\n",
    "\n",
    "#### Bedgraph for one sample of Mic\n",
    "- tool: `bedtools genomecov -bg`\n",
    "```bash\n",
    "head /home/al4225/project/multiome/step5_bedgraph_coverage/Microglia/MAP26637867.bedgraph\n",
    "chrom   start   end MAP26637867(coverage)\n",
    "chr1    10156   10157   1\n",
    "chr1    10157   10169   2\n",
    "chr1    10169   10186   3\n",
    "chr1    10186   10209   2\n",
    "chr1    10209   10210   1\n",
    "chr1    10267   10279   1\n",
    "chr1    10309   10436   2\n",
    "chr1    17339   17494   2\n",
    "chr1    87248   87394   1\n",
    "chr1    91121   91362   2\n",
    "```\n",
    "\n",
    "#### Bedgraph after merging all samples of Mic, and get PICALM region\n",
    "- tool: `bedtools unionbedg`  \n",
    "- After generating the BedGraph files for all samples, I used bedtools unionbedg to merge them. The resulting file contains chromosome, start, end, and the coverage values for multiple samples. Each row represents an interval where the coverage values are the same across samples. Also memory savings, but still maintaining detailed coverage data.\n",
    "\n",
    "- PICALM Region window:\n",
    "    - Original PICALM region expanded by 1MB on both sides:\n",
    "    - Start: chr11 84,668,727 (original: 85,668,727)\n",
    "    - End: chr11 86,780,924 (original: 85,780,924)\n",
    "    \n",
    "```bash\n",
    "zcat /home/al4225/project/multiome/step5_bedgraph_coverage/PICALM/PICALM_Microglia.bedgraph.gz |\n",
    "head | cut -f 1-6\n",
    "chrom   start             end    MAP26637867         MAP50106992      MAP61344957\n",
    "chr11   84668727        84668728        1               0               1\n",
    "chr11   84668728        84668729        1               0               1\n",
    "chr11   84668729        84668735        1               0               1\n",
    "chr11   84668735        84668736        1               0               1\n",
    "chr11   84668736        84668741        1               0               1\n",
    "chr11   84668741        84668747        1               0               1\n",
    "chr11   84668747        84668750        1               0               1\n",
    "chr11   84668750        84668751        1               0               1\n",
    "chr11   84668751        84668752        1               0               1\n",
    "```\n",
    "\n",
    "- `chrom`: Chromosome name (e.g., chr11 represents chromosome 11).\n",
    "- `start`: Start position, 0-based. It indicates the first base position in the interval.\n",
    "- `end`: End position, not inclusive (half-open interval).\n",
    "- `Sample columns` (e.g., MAP26637867, MAP50106992, MAP61344957): These columns represent the coverage value for  each specific sample. The numbers indicate the signal strength or the number of fragments covering that region.   \n",
    "\n",
    "For example, chr11 84668727 84668728 1 0 1 means that, for the region on chromosome 11 from 84668727 to 84668728, the samples MAP26637867 and MAP61344957 have a coverage of 1, while MAP50106992 has a coverage of 0.\n",
    "\n",
    "### PICALM genotype data\n",
    "- ROSMAP PLINK Genotype Data:\n",
    "    - Subsetted data covering 1153 samples (rows) and 74,512 SNPs (columns) in the region.\n",
    "    - Need to map samples to each phenotype data.\n",
    "```bash\n",
    "zcat /home/al4225/project/multiome/step5_bedgraph_coverage/PICALM_genotype/genotype/snuc_pseudo_bulk_mega.ENSG00000073921.PICALM_ROSMAP_NIA_WGS.leftnorm.bcftools_qc.plink_qc.11.tsv.gz | head -n 5| cut -f 1-4\n",
    "chr11:84668727_C_T      chr11:84668730_T_C      chr11:84668839_C_T      chr11:84668840_G_A\n",
    "MAP15387421     0       0       0\n",
    "MAP26637867     0       0       0\n",
    "MAP29629849     0       0       0\n",
    "MAP33332646     0       0       0\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
